src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpEnvironmentProxy.cs

```diff
@@ -168,18 +178,73 @@ private static Uri GetUriFromString(string value)
             {
                 return null;
             }
+            if (value.StartsWith("http://", StringComparison.OrdinalIgnoreCase))
+            {
+                value = value.Substring(7);
+            }
+            string user = null;
+            string password = null;
+            UInt16 port = 80;
+            string host = null;
+
+            // Check if there is authentication part with user and possibly password.
+            // Curl accepts raw text and that may break URI parser.
+            int separatorIndex = value.LastIndexOf('@');
+            if (separatorIndex != -1)
+            {
+                string auth = value.Substring(0, separatorIndex);
+                // The User and password may or may not be URL encoded.
+                // Curl seems to accept both. To match that,
+                // we do opportunistic decode and we use original string if it fails.
+                try
+                {
+                    auth = Uri.UnescapeDataString(auth);
+                }
+                catch { };
+
+                value = value.Substring(separatorIndex + 1);
+                separatorIndex = auth.IndexOf(':');
+                if (separatorIndex == -1)
+                {
+                    user = auth;
+                }
+                else
+                {
+                    user = auth.Substring(0, separatorIndex);
+                    password = auth.Substring(separatorIndex + 1);
+                }
+            }
 
-            if (!value.Contains("://"))
+            int ipV6AddressEnd = value.IndexOf(']');
+            separatorIndex = value.LastIndexOf(':');
+            // No ':' or it is part of IPv6 address.
+            if (separatorIndex == -1 || (ipV6AddressEnd != -1 && separatorIndex < ipV6AddressEnd))
             {
-                value = "http://" + value;
+                host = value;
+            }
+            else
+            {
+                host = value.Substring(0, separatorIndex);
+                if (!UInt16.TryParse(value.Substring(separatorIndex + 1), out port))
```

Comment:
It'd be better to avoid this Substring by using the TryParse overload that takes a span.